[case reachability_mypy_compatible]
# flags: --mypy-compatible
def foo(x: None) -> None:
    if x is None:
        return
    else:
        1()
        if x:
            reveal_type(x)

[case reachability_no_mypy_compatible]
# flags: --no-mypy-compatible
def foo(x: None, y: int | None) -> None:
    if x is None:
        return
    else:
        1()  # E: "int" not callable
        if y:
            reveal_type(y)  # N: Revealed type is "int"
            reveal_type(x)  # N: Revealed type is "None"

[case narrow_len_yoda_like]
from typing import Any, Union, Tuple

x: Union[Tuple[int, int], Tuple[int, int, int]]
if 2 == len(x):
    reveal_type(x) # N: Revealed type is "Tuple[int, int]"
else:
    reveal_type(x) # N: Revealed type is "Tuple[int, int, int]"

[case narrow_unreachable_len1]
# flags: --warn-unreachable --mypy-compatible
from typing import Tuple

x: Tuple[int, ...]
if str():
    if len(x) >= 0:
        reveal_type(x)  # N: Revealed type is "tuple[int, ...]"
    else:
        reveal_type(x)  # E: Statement is unreachable

if str():
    if len(x) <= 0:
        reveal_type(x)  # N: Revealed type is "tuple[()]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, Unpack[Tuple[int, ...]]]"

if str():
    if 0 >= len(x):
        reveal_type(x)  # N: Revealed type is "tuple[()]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, Unpack[Tuple[int, ...]]]"

if str():
    if len(x) <= 1:
        reveal_type(x)  # N: Revealed type is "tuple[()] | tuple[int]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, int, Unpack[tuple[int, ...]]]"

if str():
    if 1 >= len(x):
        reveal_type(x)  # N: Revealed type is "tuple[()] | tuple[int]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, int, Unpack[tuple[int, ...]]]"


[case narrow_unreachable_len2]
# flags: --warn-unreachable --mypy-compatible
from typing import Tuple

x: Tuple[int, ...]

if str():
    if 0 > len(x):
        reveal_type(x)  # E: Statement is unreachable
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, ...]"

if str():
    if 1 > len(x):
        reveal_type(x)  # N: Revealed type is "tuple[()]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, Unpack[Tuple[int, ...]]]"

if str():
    if 2 > len(x):
        reveal_type(x)  # N: Revealed type is "tuple[()] | tuple[int]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, int, Unpack[tuple[int, ...]]]"

if str():
    if 0 < len(x):
        reveal_type(x)  # N: Revealed type is "tuple[int, Unpack[Tuple[int, ...]]]"
    else:
        reveal_type(x)  # N: Revealed type is "tuple[()]"

if str():
    if len(x) < 0:
        reveal_type(x)  # E: Statement is unreachable
    else:
        reveal_type(x)  # N: Revealed type is "tuple[int, ...]"

[case narrowing_named_tuple_calls1]
# Like testNarrowingLenUnionOfNamedTuples, but with direct calls instead of classes
from typing import NamedTuple, Union

Point2D = NamedTuple("Point2D", [('x', int), ('y', int)])
Point3D = NamedTuple("Point3D", [('x', int), ('y', int), ('z', int)])

x: Union[Point2D, Point3D]
if len(x) == 2:
    reveal_type(x)  # N: Revealed type is "Tuple[builtins.int, builtins.int, fallback=__main__.Point2D]"
else:
    reveal_type(x)  # N: Revealed type is "Tuple[builtins.int, builtins.int, builtins.int, fallback=__main__.Point3D]"

[case narrowing_named_tuple_calls2]
# Like testNarrowingLenUnionOfNamedTuples, but with direct calls instead of classes
from typing import NamedTuple, Union
from collections import namedtuple

Point2D = namedtuple("Point2D", ['x', 'y'])
Point3D = namedtuple("Point3D", ['x', 'y', 'z'])

x: Union[Point2D, Point3D]
if len(x) == 2:
    reveal_type(x)  # N: Revealed type is "Tuple[Any, Any, fallback=__main__.Point2D]"
else:
    reveal_type(x)  # N: Revealed type is "Tuple[Any, Any, Any, fallback=__main__.Point3D]"

[case narrowing_type_of_other_side]
from typing import Any

x: Any
if int == type(x):
    reveal_type(x)  # N: Revealed type is "int"
else:
    reveal_type(x)  # N: Revealed type is "Any"

[case avoid_narrowing_of_len_with_big_unpack]
from typing import Tuple
from typing_extensions import Unpack

VarTuple = Tuple[int, Unpack[Tuple[str, ...]]]

x: VarTuple
if len(x) < 30:
    reveal_type(x) # N: Revealed type is "builtins.tuple[builtins.int, Unpack[Tuple[str, ...]]]"
else:
    reveal_type(x) # N: Revealed type is "builtins.tuple[builtins.int, Unpack[Tuple[str, ...]]]"

[case narrow_attribute_split]
class Foo:
    x: int | None

foo: Foo

foo.x + 1 # E: Unsupported operand types for + ("None" and "int") \
          # N: Left operand is of type "int | None"

if foo.x:
    foo.x + 1

[case narrow_getitem_split]
foo: list[int | None]

foo[0] + 1 # E: Unsupported operand types for + ("None" and "int") \
           # N: Left operand is of type "int | None"

if foo[0]:
    foo[0] + 1
    foo[1] + 1 # E: Unsupported operand types for + ("None" and "int") \
               # N: Left operand is of type "int | None"

[case new_type_intersection]
from typing import NewType
class A: ...
class B: ...
N = NewType('N', A)

def parse_ast(n: N) -> None:
    if isinstance(n, B):
        reveal_type(n)  # N: Revealed type is "<subclass of "N" and "B">"

[case tuple_intersection_with_getitem]
from typing import Tuple
class A:
    def foo(self) -> int: ...
class B: ...

x: Tuple[int, str]

if isinstance(x, A):
    reveal_type(x)  # N: Revealed type is "<subclass of "tuple" and "A">"
    reveal_type(x[0])  # N: Revealed type is "int"
    reveal_type(x[1])  # N: Revealed type is "str"
    x.foo()

a: A
if isinstance(a, B):
    a[0]  # E: Value of type "A" is not indexable

[case intersection_with_getitem_of_classes]
from typing import Tuple
class A:
    def __getitem__(self, other: int) -> str: ...
class B: ...

a: A
b: B
if isinstance(a, B):
    reveal_type(a[0])  # N: Revealed type is "str"

if isinstance(b, A):
    reveal_type(b[0]) # N: Revealed type is "str"

[case issubclass_and_isinstance_with_generics]
from typing import Generic, TypeVar, Type
T = TypeVar('T')

class A(Generic[T]): ...

a1: Type[A[int]]
a2: A[int]

if issubclass(a1, A):
    reveal_type(a1)  # N: Revealed type is "Type[__main__.A[int]]"

if isinstance(a2, A):
    reveal_type(a2) # N: Revealed type is "__main__.A[int]"

[case narrow_frozenset_contains]
d: frozenset[str]
key: str | None
if key in d:
    reveal_type(key)  # N: Revealed type is "str"
else:
    reveal_type(key)  # N: Revealed type is "str | None"
