[case tuple_none]
x: tuple[None]
reveal_type(x.index(0)) # N: Revealed type is "int"
reveal_type(x.__getitem__(0)) # N: Revealed type is "None"

[case tuple_methods_empty]
x: tuple[()]
reveal_type(x.index(0)) # N: Revealed type is "int"
-- reveal_type(x.__getitem__(0)) # N: Revealed type is "<nothing>"

[case tuple_methods_one]
x: tuple[str]
reveal_type(x.__getitem__(0)) # N: Revealed type is "str"
reveal_type(x.index(0)) # N: Revealed type is "int"

[case tuple_methods_multiple]
x: tuple[int, str]
reveal_type(x.__getitem__(0)) # N: Revealed type is "object"
reveal_type(x.index(0)) # N: Revealed type is "int"

